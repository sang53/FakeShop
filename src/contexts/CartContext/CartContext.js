import { createContext } from "react";

export const CartItems = createContext(null);
export const CartReducer = createContext(null);
