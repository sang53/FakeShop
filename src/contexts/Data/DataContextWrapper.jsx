import { useEffect, useState } from "react";
import Helpers from "../../Helper/Helpers";
import { Products } from "./DataContext";
import PropTypes from "prop-types";

export default function DataContextWrapper({ children }) {
  const [data, setData] = useState([]);

  useEffect(() => {
    async function getData() {
      const newData = await Helpers.getProduct();
      if (useData) setData(newData);
    }
    let useData = true;
    getData();

    return () => (useData = false);
  }, []);

  return <Products.Provider value={data}>{children}</Products.Provider>;
}

DataContextWrapper.propTypes = {
  children: PropTypes.element.isRequired,
};
