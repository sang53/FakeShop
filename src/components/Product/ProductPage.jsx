import classes from "./ProductPage.module.css";
import { use, useRef } from "react";
import { useParams } from "react-router";
import { Products } from "../../contexts/Data/DataContext";
import { CartReducer } from "../../contexts/CartContext/CartContext";
import Helper from "../../Helper/Helpers";

export default function ProductPage() {
  const { id } = useParams();
  const productData = use(Products);
  const cartDispatch = use(CartReducer);
  const inputRef = useRef(null);
  const totalRef = useRef(null);

  if (!productData.length) return <h1>Loading...</h1>;

  const product = productData.find((obj) => obj.id === +id);
  if (!product) return <h1>Error: Product Not Found (id: {id})</h1>;

  return (
    <div className={"flex-column-center pad-1rem gap-1rem " + classes.main}>
      <h2>{product.title}</h2>
      <img src={product.image} alt={product.title} />
      <div className={classes.desc}>
        {product.description
          .split(".")
          .slice(0, -1)
          .map((sent, idx) => {
            return <p key={idx}>{sent + "."}</p>;
          })}
      </div>
      <div>{Helper.padDecimal(product.price)}</div>
      <div>
        Rating: {product.rating.rate} / 5 ({product.rating.count} Votes)
      </div>
      <div>
        <label htmlFor="total">Total: </label>
        <input
          id="total"
          className="total"
          type="text"
          readOnly
          ref={totalRef}
          value={Helper.getTotal(1, product.price)}
        />
      </div>
      <div>
        <label htmlFor="add-cart">Add To Cart: </label>
        <input
          id="add-cart"
          type="number"
          min="1"
          defaultValue="1"
          ref={inputRef}
          onChange={(e) => {
            if (e.target.value <= 0)
              totalRef.current.value = Helper.getTotal(1, product.price);
            else
              totalRef.current.value = Helper.getTotal(
                e.target.value,
                product.price
              );
          }}
          onBlur={(e) => {
            if (e.target.value <= 0) inputRef.current.value = 1;
          }}
        />
        <button
          onClick={() => {
            cartDispatch({
              type: "add",
              id,
              quantity: inputRef.current.value,
            });
            inputRef.current.value = 1;
            totalRef.current.value = Helper.getTotal(1, product.price);
          }}
        >
          Add
        </button>
      </div>
    </div>
  );
}
