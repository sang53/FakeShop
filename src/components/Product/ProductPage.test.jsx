import { render, screen } from "@testing-library/react";
import { CartReducer } from "../../contexts/CartContext/CartContext";
import { describe, expect, test, vi } from "vitest";
import { Products } from "../../contexts/Data/DataContext";
import { mockProducts } from "../../test/utils/mockProducts";
import { MemoryRouter, Routes, Route } from "react-router";
import ProductPage from "./ProductPage";

function renderProduct(id = 1, loading = false) {
  const mockDispatch = vi.fn();
  const renderReturns = render(
    <CartReducer.Provider value={mockDispatch}>
      <Products.Provider value={loading ? [] : mockProducts}>
        <MemoryRouter initialEntries={[`/products/${id}`]}>
          <Routes>
            <Route path="/products/:id" element={<ProductPage />} />
          </Routes>
        </MemoryRouter>
      </Products.Provider>
    </CartReducer.Provider>
  );
  return { ...renderReturns, mockDispatch };
}

test("loading page when data is empty", () => {
  renderProduct(1, true);
  expect(screen.getByRole("heading")).toHaveTextContent(/loading/i);
});

test("error page when product id not found", () => {
  renderProduct(100);
  expect(screen.getByRole("heading")).toHaveTextContent(
    /something went wrong/i
  );
});

test("product page elements displayed as expected", () => {
  const { container } = renderProduct();
  expect(container).toMatchSnapshot();
});

describe("quantity input and total display", () => {
  test("starting values of quantity & total", () => {
    renderProduct();
    expect(screen.getByLabelText(/add/i)).toHaveValue(1);
    expect(screen.getByLabelText(/total/i)).toHaveValue("$109.95");
  });
  test("change quantity to valid quantity", async () => {
    renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change quantity without unfocusing element
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{3}");
    expect(quantityInput).toHaveValue(3);
    expect(totalInput).toHaveValue("$329.85");

    // unfocus quantity input element
    await user.click(totalInput);
    expect(quantityInput).toHaveValue(3);
    expect(totalInput).toHaveValue("$329.85");
  });
  test("change quantity to 0", async () => {
    renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change quantity without unfocusing element
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{0}");
    expect(quantityInput).toHaveValue(0);
    expect(totalInput).toHaveValue("$109.95");

    // unfocus quantity input element
    await user.click(totalInput);
    expect(quantityInput).toHaveValue(1);
    expect(totalInput).toHaveValue("$109.95");
  });
  test("change quantity to negative number", async () => {
    renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change quantity without unfocusing element
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{-}{3}");
    expect(quantityInput).toHaveValue(-3);
    expect(totalInput).toHaveValue("$109.95");

    // unfocus quantity input element
    await user.click(totalInput);
    expect(quantityInput).toHaveValue(1);
    expect(totalInput).toHaveValue("$109.95");
  });
  test("change quantity to empty string", async () => {
    renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change quantity without unfocusing element
    await user.click(quantityInput);
    await user.keyboard("{Backspace}");
    expect(quantityInput).toHaveValue(null);
    expect(totalInput).toHaveValue("$109.95");

    // unfocus quantity input element
    await user.click(totalInput);
    expect(quantityInput).toHaveValue(1);
    expect(totalInput).toHaveValue("$109.95");
  });
});

describe("add to cart function parameters", () => {
  test("add valid quantity to cart", async () => {
    const { mockDispatch } = renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const addButton = screen.getByRole("button");
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{3}");
    await user.click(addButton);
    expect(mockDispatch).toHaveBeenCalledOnce();
    expect(mockDispatch.mock.calls[0][0]).toMatchInlineSnapshot(`
      {
        "id": "1",
        "quantity": "3",
        "type": "add",
      }
    `);
    expect(quantityInput).toHaveValue(1);
    expect(screen.getByLabelText(/total/i)).toHaveValue("$109.95");
  });
  test("try add 0 quantity to cart", async () => {
    const { mockDispatch } = renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const addButton = screen.getByRole("button");
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{0}");
    await user.click(addButton);
    expect(mockDispatch).toHaveBeenCalledOnce();
    expect(mockDispatch.mock.calls[0][0]).toMatchInlineSnapshot(`
      {
        "id": "1",
        "quantity": "1",
        "type": "add",
      }
    `);
    expect(quantityInput).toHaveValue(1);
    expect(screen.getByLabelText(/total/i)).toHaveValue("$109.95");
  });
  test("try add negative quantity to cart", async () => {
    const { mockDispatch } = renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const addButton = screen.getByRole("button");
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{-}{3}");
    await user.click(addButton);
    expect(mockDispatch).toHaveBeenCalledOnce();
    expect(mockDispatch.mock.calls[0][0]).toMatchInlineSnapshot(`
      {
        "id": "1",
        "quantity": "1",
        "type": "add",
      }
    `);
    expect(quantityInput).toHaveValue(1);
    expect(screen.getByLabelText(/total/i)).toHaveValue("$109.95");
  });
  test("try add empty quantity to cart", async () => {
    const { mockDispatch } = renderProduct();
    const quantityInput = screen.getByLabelText(/add/i);
    const addButton = screen.getByRole("button");
    await user.click(quantityInput);
    await user.keyboard("{Backspace}");
    await user.click(addButton);
    expect(mockDispatch).toHaveBeenCalledOnce();
    expect(mockDispatch.mock.calls[0][0]).toMatchInlineSnapshot(`
      {
        "id": "1",
        "quantity": "1",
        "type": "add",
      }
    `);
    expect(quantityInput).toHaveValue(1);
    expect(screen.getByLabelText(/total/i)).toHaveValue("$109.95");
  });
});
