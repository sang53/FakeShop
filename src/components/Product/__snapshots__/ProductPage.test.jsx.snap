// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`product page elements displayed as expected 1`] = `
<div>
  <div
    class="flex-column-center pad-1rem gap-1rem _main_ea36a0"
  >
    <h2>
      Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
    </h2>
    <img
      alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
      src="#"
    />
    <div
      class="_desc_ea36a0"
    >
      <p>
        Your perfect pack for everyday use and walks in the forest.
      </p>
    </div>
    <div
      class="bold"
    >
      <i>
        $109.95
      </i>
    </div>
    <div>
      <b>
        Rating: 
      </b>
      3.9
       / 5 (
      120
       Votes)
    </div>
    <div>
      <label
        class="bold"
        for="total"
      >
        Total:
      </label>
      <input
        class="total"
        id="total"
        readonly=""
        type="text"
        value="$109.95"
      />
    </div>
    <div>
      <label
        class="bold"
        for="add-cart"
      >
        Add To Cart:
         
      </label>
      <input
        id="add-cart"
        min="1"
        type="number"
        value="1"
      />
      <button>
        Add
      </button>
    </div>
  </div>
</div>
`;
